geopandas/io/file.py

```diff
@@ -149,27 +149,66 @@ def convert_type(column, in_type):
     if df.empty:
         raise ValueError("Cannot write empty DataFrame to file.")
 
-    geom_type = _common_geom_type(df)
+    # Since https://github.com/Toblerity/Fiona/issues/446 resolution,
+    # Fiona allows a list of geometry types
+    geom_types = _geometry_types(df)
 
-    schema = {'geometry': geom_type, 'properties': properties}
+    schema = {'geometry': geom_types, 'properties': properties}
 
     return schema
 
 
-def _common_geom_type(df):
-    # Need to check geom_types before we write to file...
-    # Some (most?) providers expect a single geometry type:
-    # Point, LineString, or Polygon
-    geom_types = df.geometry.geom_type.unique()
+def _geometry_types(df):
+    """
+    Determine the geometry types in the GeoDataFrame for the schema.
+    """
+    if _FIONA18:
+        # Starting from Fiona 1.8, schema submitted to fiona to write a gdf
+        # can have mixed geometries:
+        # - 3D and 2D shapes can coexist in inferred schema
+        # - Shape and MultiShape types can (and must) coexist in inferred
+        #   schema
+        geom_types_2D = df[~df.geometry.has_z].geometry.geom_type.unique()
+        geom_types_2D = [gtype for gtype in geom_types_2D if gtype is not None]
+        geom_types_3D = df[df.geometry.has_z].geometry.geom_type.unique()
+        geom_types_3D = ["3D " + gtype for gtype in geom_types_3D
+                         if gtype is not None]
+        geom_types = geom_types_3D + geom_types_2D
 
-    from os.path import commonprefix
-    # use reversed geom types and commonprefix to find the common suffix,
-    # then reverse the result to get back to a geom type
-    geom_type = commonprefix([g[::-1] for g in geom_types if g])[::-1]
-    if not geom_type:
+    else:
+        # Before Fiona 1.8, schema submitted to write a gdf should have
+        # one single geometry type whenever possible:
+        # - 3D and 2D shapes cannot coexist in inferred schema
+        # - Shape and MultiShape can not coexist in inferred schema
+        geom_types = _geometry_types_back_compat(df)
+
+    if len(geom_types) == 0:
+        # Default geometry type supported by Fiona
+        # (Since https://github.com/Toblerity/Fiona/issues/446 resolution)
         return 'Unknown'
```

Comment:
we could replace by `geom_types = 'Unknown'` to avoid multiple return statements ?

Reply:
@jorisvandenbossche what do you think ?

Reply:
Sorry, forgot to answer. I don't really care, but since the current one is actually shorter, I don't mind the multiple returns